internal/pkg/coordinator/monitor.go

```diff
@@ -0,0 +1,360 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package coordinator
+
+import (
+	"context"
+	"encoding/json"
+	"net"
+	"os"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+
+	"fleet/internal/pkg/bulk"
+	"fleet/internal/pkg/config"
+	"fleet/internal/pkg/dl"
+	"fleet/internal/pkg/es"
+	"fleet/internal/pkg/model"
+	"fleet/internal/pkg/monitor"
+)
+
+const (
+	defaultCheckInterval           = 20 * time.Second // check for valid leaders every 20 seconds
+	defaultLeaderInterval          = 30 * time.Second // become leader for at least 30 seconds
+	defaultMetadataInterval        = 5 * time.Minute  // update metadata every 5 minutes
+	defaultCoordinatorRestartDelay = 5 * time.Second  // delay in restarting coordinator on failure
+)
+
+// Monitor monitors the leader election of policies and routes managed policies to the coordinator.
+type Monitor interface {
+	// Run runs the monitor.
+	Run(context.Context) error
+}
+
+type policyT struct {
+	id        string
+	cord      Coordinator
+	canceller context.CancelFunc
+}
+
+type monitorT struct {
+	log zerolog.Logger
+
+	bulker  bulk.Bulk
+	monitor monitor.Monitor
+	factory Factory
+
+	fleet         config.Fleet
+	version       string
+	agentMetadata model.AgentMetadata
+	hostMetadata  model.HostMetadata
+
+	checkInterval     time.Duration
+	leaderInterval    time.Duration
+	metadataInterval  time.Duration
+	coordRestartDelay time.Duration
+
+	policies map[string]policyT
+}
+
+// NewMonitor creates a new coordinator policy monitor.
+func NewMonitor(fleet config.Fleet, version string, bulker bulk.Bulk, monitor monitor.Monitor, factory Factory) Monitor {
+	return &monitorT{
+		log:               log.With().Str("ctx", "policy leader manager").Logger(),
```

Comment:
Do we plan to use ECS logger?

Reply:
We are currently using the zerolog, which works nicely.

Reply:
No problem at the moment but we should make sure that fleet-server has ECS / JSON formatted logs when Elastic Agent (FB) starts to collect it to no transformation is needed.